### ISSUE TITLE
[AArch64/SVE][Inductor] test_vec_compare_op_cpu_only consumes excessive CPU and memory in scheduler dependency analysis

### ISSUE THREAD

**Issue by AnthonyBarbier** [pytorch/pytorch#192498]
### 🐛 Describe the bug

This test hangs for over an hour in our CI

Codex produced the following analysis: 

This appears to be an upstream Inductor compile-time expression explosion on AArch64/SVE.

# Reproduction

Running CPUReproTests::test_vec_compare_op_cpu_only from test/inductor/test_cpu_repro.py

# Observed behaviour

  After approximately two minutes:

- No test progress or result.
- Approximately 102% CPU: one core continuously occupied.
- RSS grew to approximately 3.1 GB.
- Process remained runnable, not sleeping or waiting on a lock.
- I stopped it manually to avoid further memory growth.

A standalone reproduction containing only the first float32 case behaved identically:

- Still compiling after 1 minute 52 seconds.
- Approximately 101% CPU.
- Approximately 2.87 GB RSS.

This proves the first float32 compilation alone is sufficient.

# Compiler stack

Early stack samples showed Inductor constructing and tracing the fused pointwise loop body:

   torch/_inductor/index_propagation.py:_default
  torch/_inductor/lowering.py:inner_fn  # repeated deeply
  torch/_inductor/loop_body.py
  torch/_inductor/ir.py:simplify_and_reorder
  torch/_inductor/scheduler.py:_compute_attrs

  Later samples consistently showed dependency-expression processing:

  sympy/core/function.py:expand
  torch/_inductor/dependencies.py:_normalize
  torch/_inductor/dependencies.py:canonicalize
  torch/_inductor/dependencies.py:load
  torch/_inductor/dependencies.py:extract_read_writes
  torch/_inductor/scheduler.py:_compute_attrs

  The standalone reproduction showed the same hotspot, including:

  torch/_inductor/codegen/common.py:index_prevent_reordering
  torch/_inductor/dependencies.py:_normalize
  torch/_inductor/dependencies.py:canonicalize
  torch/_inductor/dependencies.py:extract_read_writes

  It never reached C++ compilation or execution.

# Likely cause

  The test builds a chain of 12 comparisons and torch.where operations. Each new operation references the preceding expression multiple times.

  The stack traces suggest Inductor fuses these operations, recursively inlines the previous pointwise expression, and then expands/ canonicalizes the resulting dependency expressions. This likely causes exponential expression growth in scheduler analysis.

  That diagnosis is an inference from the repeated lowering.inner_fn frames, SymPy expansion, sustained CPU use, and rapid memory growth.

# Environment

  Architecture: aarch64
  CPU: AWS Graviton4 / Arm Neoverse-V2, 16 cores
  Vector ISA selected by Inductor: VecSVE
  Kernel: Linux 7.0.0-1010-aws
  Python: 3.12.12
  PyTorch: 2.14.0.dev20260719+cpu
  Pinned PyTorch source: d56735d8ac8cca8062e7c5df10a02f9587d63b98
  Wheel git version: 194fcb011f035ee09da225a40f5da5df0930d27c
  SymPy: 1.14.0

  The wheel commit is the July 19 nightly release built from the pinned d56735d8 source commit.

# Additional evidence

- Test timed out in our CI after 3,600 seconds.
- The same test passed in upstream x86 CPU CI.

Expected behaviour: compilation should complete in a reasonable time with bounded memory usage, comparable to x86 CPU.

### Error logs

No output: just hangs

### Versions

Architecture: aarch64
CPU: AWS Graviton4 / Arm Neoverse-V2, 16 cores
Vector ISA selected by Inductor: VecSVE
Kernel: Linux 7.0.0-1010-aws
Python: 3.12.12
PyTorch: 2.14.0.dev20260719+cpu
Pinned PyTorch source: https://github.com/pytorch/pytorch/commit/d56735d8ac8cca8062e7c5df10a02f9587d63b98
Wheel git version: https://github.com/pytorch/pytorch/commit/194fcb011f035ee09da225a40f5da5df0930d27c
SymPy: 1.14.0

The wheel commit is the July 19 nightly release built from the pinned https://github.com/pytorch/pytorch/commit/d56735d8ac8cca8062e7c5df10a02f9587d63b98 source commit.

cc @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka @masnesral @aditvenk

**Comment by aditew01** [comment]
cc: @robert-hardwick

**Comment by robert-hardwick** [comment]
We have done a bisect to find the commit that introduced the regression

  The introducing change is very likely commit 9eb9f605f32, PR #184411, “Tune CPU realization thresholds.”

  It raised the CPU realization threshold from 30 to 50, allowing this expression chain to remain fused until it explodes.

  Causal validation on the latest nightly:

```
   Configuration            Stage-10 LoopBody    Full stage-12 result
  ━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━━
   Current threshold 50         239,442 nodes    >240s, 3.36 GiB
  ───────────────────────  ───────────────────  ──────────────────────
   Previous threshold 30         26,606 nodes    10.46s, 435 MiB
```

  So restoring the old threshold completely fixes the reproducer:

```
  from torch._inductor import config

  with config.patch(realize_opcount_threshold=30):
      compiled = torch.compile(fn)
      result = compiled(x)
```

### Additional debug

Here, one “stage” means one chained comparison plus torch.where, where the result feeds the next pair.

```
  stages    Frontend FX nodes   LoopBody FX nodes
  6                     20                  2,958
  8                     26                 26,606
  10                    32                 239,442
  11                    35                 718,323
```

  The frontend grows linearly, but LoopBody._init_with_tracing grows ~3× per added pair because the previous expression is evaluated in the condition and both where branches.

  py-spy shows the resulting time in nested lowering.inner_fn, simplify_and_reorder, _compute_attrs, and dependency analysis. This suggests the realization heuristic underestimates expanded producer fanout.

@jansel @aditew01 you might be best place to propose a fix for this?

**Comment by aditew01** [comment]
@claude triage this
